### PR TITLE
Loader

### PR DESCRIPTION
**Pull request by idoc07**

### CODE REVIEW

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Loader extends SubsystemBase {
+    private WPI_TalonSRX motor;
+
+    private LoaderConstants.LoaderState currentState;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: static import

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    private final static Loader INSTANCE = new Loader();
+
+    public static Loader getInstance() {
+        return INSTANCE;
+    }
+
+    private Loader() {
+        motor = LoaderConstants.MOTOR;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: move initialization to definition

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public void setState(LoaderConstants.LoaderState state) {
+        motor.set(ControlMode.PercentOutput, state.voltage / LoaderConstants.VOLTAGE_COMPENSATION);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: set `currentState`...

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class LoaderConstants {
+    private static final int MOTOR_ID = 0;
+    protected static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: move to bottom

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    protected static final double
+            LOAD_VOLTAGE = 6,
+            EJECT_VOLTAGE = -6;
+    protected static final double VOLTAGE_COMPENSATION = 10.5;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: change all protected fields to default

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    public static Loader getInstance() {
+        return INSTANCE;
+    }
+
+    private Loader() {
+
+    }
+
+    public void setState(LoaderState state) {
+        motor.set(ControlMode.PercentOutput, state.voltage / LoaderConstants.VOLTAGE_COMPENSATION);
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: `ControlMode.PercentOutput` is unnecessary as that is the default

Comment (lycaon7) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: `.set` takes a percent between 1 and -1 not a voltage. there is a `.setVoltage` however unless @laviRZ told you to use voltage and not a percent then voltage is unnecessary here.

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class LoaderConstants {
+    private static final int MOTOR_ID = 0;
+    static final double
+            LOAD_VOLTAGE = 6,
+            EJECT_VOLTAGE = -6;
+    static final double VOLTAGE_COMPENSATION = 10.5;
+    static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java:13]: put above the voltages under the id

**Review by lycaon7 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import static frc.trigon.robot.subsystems.loader.LoaderConstants.LoaderState;
+
+public class Loader extends SubsystemBase {
+    private WPI_TalonSRX motor = LoaderConstants.MOTOR;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: final

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class LoaderConstants {
+    private static final int MOTOR_ID = 0;
+    static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
+    static final double
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: private

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class LoaderConstants {
+    private static final int MOTOR_ID = 0;
+    static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: move to the bottom of the constants

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    static final double
+            LOAD_VOLTAGE = 6,
+            EJECT_VOLTAGE = -6;
+    static final double VOLTAGE_COMPENSATION = 10.5;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: voltage compensation is not enabled

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class LoaderConstants {
+    private static final double
+            LOAD_VOLTAGE = 6,
+            EJECT_VOLTAGE = -6;
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java:9]: This should be power not voltage

**Review by lycaon7 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+import static frc.trigon.robot.subsystems.loader.LoaderConstants.LoaderState;
+
+public class Loader extends SubsystemBase {
+    private final static Loader INSTANCE = new Loader();
+    private final WPI_TalonSRX motor = LoaderConstants.MOTOR;
+    private LoaderState currentState;
+
+    private Loader() {
+
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: extra newline

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    public static Loader getInstance() {
+        return INSTANCE;
+    }
+
+    public LoaderState getState() {
+        return currentState;
+    }
+
+    public void setState(LoaderState state) {
+        motor.set(state.power / LoaderConstants.VOLTAGE_COMPENSATION);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: extract to a function inside `utilities.Conversions` called `voltageToCompensatedPower`

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 15]
```diff
@@ -0,0 +1,29 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class LoaderConstants {
+    static final double VOLTAGE_COMPENSATION = 10.5;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: rename to voltage compensation saturation

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    private static final double
+            LOAD_POWER = 6,
+            EJECT_POWER = -6;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: this is voltage, not power

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java (hunk 15 above)
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: add an inversion flag

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import frc.trigon.robot.utilities.Conversions;
+
+public class LoaderConstants {
+    private static final double
+            LOAD_VOLTAGE = 6,
+            EJECT_VOLTAGE = -6;
+    private static final boolean IS_INVERTED = false;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: rename to motor inverted

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package frc.trigon.robot.utilities;
+
+public class Conversions {
+    public static final double VOLTAGE_SATURATION = 10.5;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: rename to voltage compensation saturation

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 19]
```diff
@@ -0,0 +1,9 @@
...
+
+    public static double voltageToCompensatedVoltage(double voltage) {
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: to compensated power, not compensated voltage

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+    private Loader() {
+    }
+
+    public static Loader getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Gets the state of the Loader
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: redundant

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+     *
+     * @return the current LoaderState
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: The current state of the loader. There's no need to specify the return type in the doc.

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    public LoaderState getState() {
+        return currentState;
+    }
+
+    /**
+     * Sets the state of the Loader
+     *
+     * @param state the wanted LoaderState
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: don't use the type of the parameter. "the wanted state for the loader" is a good description

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 23]
```diff
@@ -0,0 +1,15 @@
+package frc.trigon.robot.utilities;
+
+public class Conversions {
+    public static final double VOLTAGE_COMPENSATION_SATURATION = 10.5;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: this is motor dependent

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 24]
```diff
@@ -0,0 +1,20 @@
+package frc.trigon.robot.utilities;
+
+public class Conversions {
+
+    /**
+     * Converts voltage to compensated power
+     *
+     * Talon FX/SRX and Victor SPX can be configured to adjust their outputs in response to the battery
+     * voltage measurement (in all control modes). Use the voltage compensation saturation config to
+     * determine what voltage represents 100% output.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: I don't want this copypasta here

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+     * The compensated power is the voltage divided by the saturation of the compensation
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: I don't want the calculation, I want to know what saturation and compensation means.

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+
+public class Loader extends SubsystemBase {
+    private final static Loader INSTANCE = new Loader();
+    private final WPI_TalonSRX motor = LoaderConstants.MOTOR;
+    private LoaderState currentState;
+
+    private Loader() {
+    }
+
+    public static Loader getInstance() {
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java:18]: Add Java Docs

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 27]
```diff
@@ -0,0 +1,32 @@
+package frc.trigon.robot.subsystems.loader;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import frc.trigon.robot.utilities.Conversions;
+
+public class LoaderConstants {
+    private static final double
+            LOAD_VOLTAGE = 6,
+            EJECT_VOLTAGE = -6;
+    public static final double VOLTAGE_COMPENSATION_SATURATION = 10.5;
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java:10]: This is global. put the compensation as constant in conversions

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+
+    static {
+        MOTOR.setInverted(MOTOR_INVERTED);
+        MOTOR.enableVoltageCompensation(true);
+        MOTOR.configVoltageCompSaturation(VOLTAGE_COMPENSATION_SATURATION);
+    }
+
+    public enum LoaderState {
+        LOAD(LOAD_VOLTAGE),
+        OFF(0),
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java]: put `OFF(0)` below eject

**Review by lycaon7 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java (hunk 26 above)
Comment (shay124578) [reply src/main/java/frc/trigon/robot/subsystems/loader/Loader.java:18]: he doesn't need to do that

File: src/main/java/frc/trigon/robot/utilities/Conversions.java (hunk 24 above)
Comment (idoc07) [reply src/main/java/frc/trigon/robot/utilities/Conversions.java]: bruh

File: src/main/java/frc/trigon/robot/utilities/Conversions.java (hunk 25 above)
Comment (idoc07) [reply src/main/java/frc/trigon/robot/utilities/Conversions.java]: bruh

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java (hunk 26 above)
Comment (idoc07) [reply src/main/java/frc/trigon/robot/subsystems/loader/Loader.java:18]: do I need or not?

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * @return the current state of the loader
+     */
+    public LoaderState getState() {
+        return currentState;
+    }
+
+    /**
+     * Sets the state of the loader
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: Dot at the end

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+     *
+     * @param state the wanted state for the loader
+     */
+    public void setState(LoaderState state) {
+        motor.set(Conversions.voltageToCompensatedPower(state.voltage, LoaderConstants.VOLTAGE_COMPENSATION_SATURATION));
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: store in a variable before this line

File: src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+    private static final double
+            LOAD_VOLTAGE = 6,
+            EJECT_VOLTAGE = -6;
+    public static final double VOLTAGE_COMPENSATION_SATURATION = 10.5;
+    private static final boolean MOTOR_INVERTED = false;
+    private static final int MOTOR_ID = 0;
+    static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
+
+    static {
+        MOTOR.setInverted(MOTOR_INVERTED);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/loader/LoaderConstants.java:19]: separate the two blocks of code here with a newline

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 32]
```diff
@@ -0,0 +1,20 @@
+package frc.trigon.robot.utilities;
+
+public class Conversions {
+
+    /**
+     * Converts voltage to compensated power
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: Dot at the end

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+     *
+     * The voltage compensation saturation will determine what voltage represents 100% output.
+     *
+     * The compensated power is the power resulting from turning the voltage off and on without stopping.
+     * It's basically the voltage divided by the saturation of the compensation.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java:11]: Not good enough
You need to explain in general the point of voltage compensation.
Then for each parameter explain its meaning.

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 34]
```diff
@@ -0,0 +1,20 @@
...
+     *
+     * @param voltage the voltage
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: the voltage of what

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 35]
```diff
@@ -0,0 +1,20 @@
...
+     * @param voltageCompensationSaturation the saturation of the compensation
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: the saturation of the compensation?

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 36]
```diff
@@ -0,0 +1,20 @@
...
+     * @return the compensated power
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java:15]: compensated for what?

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/Conversions.java (hunk 35 above)
Comment (idoc07) [reply src/main/java/frc/trigon/robot/utilities/Conversions.java:14]: yes

File: src/main/java/frc/trigon/robot/utilities/Conversions.java (hunk 36 above)
Comment (idoc07) [reply src/main/java/frc/trigon/robot/utilities/Conversions.java:15]: Any problem???

**Comment by idoc07** [comment]
I'm retiring, goodbye.

File: src/main/java/frc/trigon/robot/subsystems/loader/Loader.java (hunk 30 above)
Comment (idoc07) [reply src/main/java/frc/trigon/robot/subsystems/loader/Loader.java]: done

File: src/main/java/frc/trigon/robot/utilities/Conversions.java (hunk 33 above)
Comment (idoc07) [reply src/main/java/frc/trigon/robot/utilities/Conversions.java:11]: IDK

**Pull request closed without merging**